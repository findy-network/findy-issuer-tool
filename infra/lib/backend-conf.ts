import * as cdk from "@aws-cdk/core";
import { Bucket, BlockPublicAccess, IBucket } from "@aws-cdk/aws-s3";
import { BucketDeployment, Source } from "@aws-cdk/aws-s3-deployment";
import { IHostedZone } from "@aws-cdk/aws-route53";

import { writeFileSync, rmSync } from "fs";
import { execSync } from "child_process";

export interface IssuerToolBackendConfStackProps extends cdk.StackProps {
  prod: boolean;
  ecrURL: string;
}

export class IssuerToolBackendConfStack extends cdk.Stack {
  public readonly bucket: IBucket;
  constructor(
    scope: cdk.Construct,
    id: string,
    zone: IHostedZone,
    props: IssuerToolBackendConfStackProps
  ) {
    super(scope, `${id}-conf`, props);
    const { prod, ecrURL } = props;
    const filesPath = "./.secrets";
    const confBucket = new Bucket(this, `${id}-conf-bucket`, {
      bucketName: `${id}-conf-bucket`,
      removalPolicy: prod
        ? cdk.RemovalPolicy.RETAIN
        : cdk.RemovalPolicy.DESTROY,
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
    });

    const dockerRunJson = {
      AWSEBDockerrunVersion: "1",
      Image: {
        Name: ecrURL,
        Update: "true",
      },
      Ports: [
        {
          ContainerPort: 3001,
          HostPort: 3001,
        },
      ],
    };
    const dockerRunPath = `${filesPath}/Dockerrun.aws.json`;
    writeFileSync(dockerRunPath, JSON.stringify(dockerRunJson));
    const out1 = execSync(
      `zip -r -j ${filesPath}/Dockerrun.zip ${dockerRunPath}`
    );
    console.log(out1.toString());
    const out2 = execSync(`zip -r ${filesPath}/Dockerrun.zip .ebextensions`);
    console.log(out2.toString());

    rmSync(dockerRunPath);

    new BucketDeployment(this, `${id}-bucket-deployment`, {
      sources: [Source.asset(filesPath)],
      destinationBucket: confBucket,
    });

    this.bucket = confBucket;
  }
}
