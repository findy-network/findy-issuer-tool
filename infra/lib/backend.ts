import * as cdk from "@aws-cdk/core";
import { IBucket } from "@aws-cdk/aws-s3";
import { IHostedZone } from "@aws-cdk/aws-route53";

import {
  CfnApplication,
  CfnApplicationVersion,
  CfnEnvironment,
} from "@aws-cdk/aws-elasticbeanstalk";
import {
  CfnInstanceProfile,
  Effect,
  ManagedPolicy,
  PolicyDocument,
  PolicyStatement,
  Role,
  ServicePrincipal,
} from "@aws-cdk/aws-iam";
import { CfnTable, Table } from "@aws-cdk/aws-dynamodb";
import { Secret } from "@aws-cdk/aws-secretsmanager";

export interface IssuerToolBackendStackProps extends cdk.StackProps {
  prod: boolean;
  ecrURL: string;
  bucket: IBucket;
}

export class IssuerToolBackendStack extends cdk.Stack {
  public readonly envDomain: string;
  public readonly envName: string;
  public readonly appName: string;
  constructor(
    scope: cdk.Construct,
    id: string,
    props: IssuerToolBackendStackProps
  ) {
    super(scope, `${id}-backend`, props);
    const { bucket } = props;

    const app = new CfnApplication(this, `${id}-app`, {
      applicationName: `${id}-app`,
      resourceLifecycleConfig: {
        serviceRole: `arn:aws:iam::${this.account}:role/aws-service-role/elasticbeanstalk.amazonaws.com/AWSServiceRoleForElasticBeanstalk`,
        versionLifecycleConfig: {
          maxAgeRule: {
            enabled: true,
            maxAgeInDays: 3,
          },
        },
      },
    });
    const version = new CfnApplicationVersion(this, `${id}-app-version`, {
      applicationName: app.applicationName!,
      sourceBundle: {
        s3Bucket: bucket.bucketName,
        s3Key: "Dockerrun.zip",
      },
    });
    version.addDependsOn(app);

    const table = new CfnTable(this, `${id}-table`, {
      tableName: "issuerToolData",
      attributeDefinitions: [
        { attributeName: "type", attributeType: "S" },
        { attributeName: "id", attributeType: "S" },
      ],
      keySchema: [
        { attributeName: "type", keyType: "HASH" },
        { attributeName: "id", keyType: "RANGE" },
      ],
      provisionedThroughput: {
        readCapacityUnits: 5,
        writeCapacityUnits: 5,
      },
    });

    const policy = ManagedPolicy.fromAwsManagedPolicyName(
      "AmazonEC2ContainerRegistryReadOnly"
    );
    const secret = Secret.fromSecretNameV2(
      this,
      `issuer-tool-secret-name`,
      "issuer-tool"
    );
    const policyDoc = new PolicyDocument({
      statements: [
        new PolicyStatement({
          sid: "AllowDynamoDb",
          effect: Effect.ALLOW,
          actions: ["dynamodb:*"],
          resources: [table.attrArn],
        }),
        new PolicyStatement({
          sid: "AllowCloudWatch",
          effect: Effect.ALLOW,
          actions: [
            "logs:CreateLogGroup",
            "logs:CreateLogStream",
            "logs:GetLogEvents",
            "logs:PutLogEvents",
            "logs:DescribeLogGroups",
            "logs:DescribeLogStreams",
            "logs:PutRetentionPolicy",
          ],
          resources: ["*"],
        }),
        new PolicyStatement({
          sid: "AllowS3Read",
          effect: Effect.ALLOW,
          actions: ["s3:GetObject"],
          resources: [`${bucket.bucketArn}/*`],
        }),
        new PolicyStatement({
          sid: "AllowSecretRead",
          effect: Effect.ALLOW,
          actions: ["secretsmanager:GetSecretValue"],
          resources: [`${secret.secretArn}*`],
        }),
      ],
      // TODO: do we need allow security group?
    });
    const roleName = `${id}-instance-profile-role`;
    const instanceProfileRole = new Role(this, roleName, {
      roleName,
      assumedBy: new ServicePrincipal("ec2.amazonaws.com"),
      managedPolicies: [policy],
      inlinePolicies: { "issuer-tool-policy": policyDoc },
    });
    instanceProfileRole.assumeRolePolicy?.addStatements(
      new PolicyStatement({
        effect: Effect.ALLOW,
        principals: [new ServicePrincipal("ec2.amazonaws.com")],
        actions: ["sts:AssumeRole"],
      })
    );

    const instanceProfile = new CfnInstanceProfile(
      this,
      "issuer-tool-instance-profile",
      {
        instanceProfileName: roleName,
        path: "/",
        roles: [instanceProfileRole.roleName],
      }
    );

    const env = new CfnEnvironment(this, `${id}-environment`, {
      applicationName: version.applicationName,
      environmentName: version.applicationName,
      versionLabel: version.ref,
      cnamePrefix: "issuer-tool",
      solutionStackName:
        "64bit Amazon Linux 2018.03 v2.17.3 running Docker 20.10.7",
      optionSettings: [
        {
          namespace: "aws:autoscaling:launchconfiguration",
          optionName: "IamInstanceProfile",
          value: instanceProfile.ref,
        },
        {
          namespace: "aws:elasticbeanstalk:environment",
          optionName: "EnvironmentType",
          value: "SingleInstance",
        },
        {
          namespace: "aws:elasticbeanstalk:environment",
          optionName: "EnvironmentType",
          value: "SingleInstance",
        },
        {
          namespace: "aws:autoscaling:asg",
          optionName: "MinSize",
          value: "1",
        },
        {
          namespace: "aws:autoscaling:asg",
          optionName: "MaxSize",
          value: "1",
        },
      ],
    });
    this.envDomain = `${env.cnamePrefix}.${this.region}.elasticbeanstalk.com`;
    this.envName = env.environmentName!;
    this.appName = version.applicationName;
  }
}
