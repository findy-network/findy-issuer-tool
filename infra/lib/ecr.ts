import * as cdk from "@aws-cdk/core";
import { Repository } from "@aws-cdk/aws-ecr";
import { CfnOutput } from "@aws-cdk/core";

export interface IssuerToolEcrStackProps extends cdk.StackProps {
  prod: boolean;
}

export class IssuerToolEcrStack extends cdk.Stack {
  public readonly ecrURL: CfnOutput;
  public readonly imageName: string;
  constructor(
    scope: cdk.Construct,
    id: string,
    props: IssuerToolEcrStackProps
  ) {
    super(scope, `${id}-ecr`, props);
    const { prod } = props;
    const repository = new Repository(this, `${id}-repository`, {
      imageScanOnPush: true,
      removalPolicy: prod
        ? cdk.RemovalPolicy.RETAIN
        : cdk.RemovalPolicy.DESTROY,
    });
    repository.addLifecycleRule({
      description: "Retain max 3 images",
      maxImageCount: 3,
    });
    this.ecrURL = new CfnOutput(this, `${id}-ecr-url`, {
      value: repository.repositoryUriForDigest(),
    });
    this.imageName = repository.repositoryName;
  }
}
