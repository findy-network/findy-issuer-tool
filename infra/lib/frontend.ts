import * as cdk from "@aws-cdk/core";
import {
  OriginAccessIdentity,
  OriginProtocolPolicy,
} from "@aws-cdk/aws-cloudfront";
import { Bucket, BlockPublicAccess, IBucket } from "@aws-cdk/aws-s3";
import { PolicyStatement } from "@aws-cdk/aws-iam";
import { DnsValidatedCertificate } from "@aws-cdk/aws-certificatemanager";
import {
  CloudFrontWebDistribution,
  SSLMethod,
  SecurityPolicyProtocol,
  CloudFrontAllowedMethods,
} from "@aws-cdk/aws-cloudfront";
import { ARecord, RecordTarget } from "@aws-cdk/aws-route53";
import { CloudFrontTarget } from "@aws-cdk/aws-route53-targets/lib";
import { IHostedZone } from "@aws-cdk/aws-route53";

export interface IssuerToolFrontendStackProps extends cdk.StackProps {
  bucketName: string;
  prod: boolean;
  domainName: string;
  apiDomain: string;
  apiPaths: string[];
}

export class IssuerToolFrontendStack extends cdk.Stack {
  public readonly bucket: IBucket;
  constructor(
    scope: cdk.Construct,
    id: string,
    zone: IHostedZone,
    props: IssuerToolFrontendStackProps
  ) {
    super(scope, `${id}-frontend`, props);

    const { bucketName, prod } = props;

    const bucket = new Bucket(this, `${id}-bucket`, {
      bucketName: bucketName,
      websiteIndexDocument: "index.html",
      websiteErrorDocument: "index.html",
      removalPolicy: prod
        ? cdk.RemovalPolicy.RETAIN
        : cdk.RemovalPolicy.DESTROY,
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
    });

    const bucketOriginAccessIdentity = new OriginAccessIdentity(
      this,
      `${id}-origin-access-identity`,
      {
        comment: `Access bucket ${bucketName} only from Cloudfront`,
      }
    );
    const policyStatement = new PolicyStatement();
    policyStatement.addActions("s3:GetObject*");
    policyStatement.addResources(`${bucket.bucketArn}/*`);
    policyStatement.addCanonicalUserPrincipal(
      bucketOriginAccessIdentity.cloudFrontOriginAccessIdentityS3CanonicalUserId
    );
    bucket.addToResourcePolicy(policyStatement);
    const listPolicyStatement = new PolicyStatement();
    listPolicyStatement.addActions("s3:ListBucket");
    listPolicyStatement.addResources(bucket.bucketArn);
    listPolicyStatement.addCanonicalUserPrincipal(
      bucketOriginAccessIdentity.cloudFrontOriginAccessIdentityS3CanonicalUserId
    );
    bucket.addToResourcePolicy(listPolicyStatement);

    const s3Origin = {
      s3BucketSource: bucket,
      originAccessIdentity: bucketOriginAccessIdentity,
    };

    const { domainName, apiDomain, apiPaths } = props!;

    // To use an ACM certificate with Amazon CloudFront, you must request or import the certificate
    // in the US East (N. Virginia) region. ACM certificates in this region that are associated
    // with a CloudFront distribution are distributed to all the geographic locations configured for that distribution.
    const certificateArn = new DnsValidatedCertificate(
      this,
      `${id}-certificate`,
      {
        domainName,
        hostedZone: zone,
        region: "us-east-1",
      }
    ).certificateArn;

    const distribution = new CloudFrontWebDistribution(
      this,
      `${id}-distribution`,
      {
        aliasConfiguration: {
          acmCertRef: certificateArn,
          names: [domainName],
          sslMethod: SSLMethod.SNI,
          securityPolicy: SecurityPolicyProtocol.TLS_V1_2_2019,
        },
        errorConfigurations: [
          {
            errorCode: 404,
            responsePagePath: "/index.html",
            responseCode: 200,
            errorCachingMinTtl: 0,
          },
        ],
        originConfigs: [
          {
            s3OriginSource: s3Origin,
            behaviors: [
              {
                isDefaultBehavior: true,
              },
              {
                pathPattern: "/index.html",
                maxTtl: cdk.Duration.seconds(0),
                minTtl: cdk.Duration.seconds(0),
                defaultTtl: cdk.Duration.seconds(0),
              },
            ],
          },
          {
            customOriginSource: {
              domainName: apiDomain,
              originProtocolPolicy: OriginProtocolPolicy.HTTP_ONLY,
              originReadTimeout: cdk.Duration.seconds(60),
              originKeepaliveTimeout: cdk.Duration.seconds(60),
            },
            behaviors: apiPaths.map((item: string) => ({
              pathPattern: item,
              allowedMethods: CloudFrontAllowedMethods.ALL,
              forwardedValues: {
                cookies: {
                  forward: "all",
                },
                headers: ["*"],
                queryString: true,
              },
            })),
          },
        ],
      }
    );

    new ARecord(this, `${id}-a-record`, {
      recordName: domainName,
      target: RecordTarget.fromAlias(new CloudFrontTarget(distribution)),
      zone,
    });
    this.bucket = bucket;
  }
}
