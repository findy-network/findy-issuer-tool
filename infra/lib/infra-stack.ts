import {
  Stack,
  StackProps,
  RemovalPolicy,
  Duration,
  CfnOutput,
} from "aws-cdk-lib";
import { Construct } from "constructs";
import {
  CfnApplication,
  CfnApplicationVersion,
  CfnEnvironment,
} from "aws-cdk-lib/aws-elasticbeanstalk";
import { CfnTable } from "aws-cdk-lib/aws-dynamodb";
import { IBucket, Bucket, BlockPublicAccess } from "aws-cdk-lib/aws-s3";
import { BucketDeployment, Source } from "aws-cdk-lib/aws-s3-deployment";
import { Secret } from "aws-cdk-lib/aws-secretsmanager";
import {
  CfnInstanceProfile,
  Effect,
  PolicyDocument,
  PolicyStatement,
  Role,
  ServicePrincipal,
} from "aws-cdk-lib/aws-iam";
import {
  OriginAccessIdentity,
  OriginProtocolPolicy,
  CloudFrontWebDistribution,
  SSLMethod,
  SecurityPolicyProtocol,
  CloudFrontAllowedMethods,
} from "aws-cdk-lib/aws-cloudfront";
import { DnsValidatedCertificate } from "aws-cdk-lib/aws-certificatemanager";
import { ARecord, RecordTarget, HostedZone } from "aws-cdk-lib/aws-route53";
import { CloudFrontTarget } from "aws-cdk-lib/aws-route53-targets";

import { Frontend } from "./frontend";
import { Backend } from "./backend";

export class InfraStack extends Stack {
  public readonly appName: CfnOutput;
  public readonly envName: CfnOutput;
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const backend = new Backend(this, `${id}-backend`, {
      account: this.account,
    });

    this.envName = new CfnOutput(this, "EnvName", {
      value: backend.envName,
    });
    this.appName = new CfnOutput(this, "AppName", {
      value: backend.appName,
    });

    const apiPaths = [
      "/auth/*",
      "/issuer*",
      "/user*",
      "/ledger*",
      "/events/*",
      "/connections*",
      "/create/*",
      "/pairwise/*",
      "/creds/*",
      "/ftn/*",
    ];
    const backendUrl = `${backend.cnamePrefix}.${this.region}.elasticbeanstalk.com`;
    new Frontend(this, `${id}-frontend`, {
      rootDomainName: process.env.DOMAIN_NAME || "example.com",
      appDomainPrefix: process.env.SUB_DOMAIN_NAME || "example",
      apiPaths,
      backendUrl,
    });
  }
}
