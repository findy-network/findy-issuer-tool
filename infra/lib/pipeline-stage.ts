import { InfraStack } from "./infra-stack";
import { Stage, StageProps, CfnOutput } from "aws-cdk-lib";
import { Construct } from "constructs";

export class InfraPipelineStage extends Stage {
  public readonly appName: CfnOutput;
  public readonly envName: CfnOutput;
  constructor(scope: Construct, id: string, props?: StageProps) {
    super(scope, id, props);

    const service = new InfraStack(this, "FindyIssuerToolInfraStack", props);
    this.appName = service.appName;
    this.envName = service.envName;
  }
}
