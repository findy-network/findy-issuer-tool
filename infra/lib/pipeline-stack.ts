import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import {
  CodeBuildStep,
  CodePipeline,
  CodePipelineSource,
} from "aws-cdk-lib/pipelines";
import { aws_codebuild as codebuild, aws_logs as logs } from "aws-cdk-lib";
import { StringParameter } from "aws-cdk-lib/aws-ssm";
import { PolicyStatement, Role, ServicePrincipal } from "aws-cdk-lib/aws-iam";
import { Bucket, BlockPublicAccess, IBucket } from "aws-cdk-lib/aws-s3";
import { BucketDeployment, Source } from "aws-cdk-lib/aws-s3-deployment";
import { writeFileSync, mkdirSync } from "fs";
import { execSync } from "child_process";

import { InfraPipelineStage } from "./pipeline-stage";

interface InfraPipelineProperties extends cdk.StackProps {
  skipConfigCopy?: boolean;
}

const environmentVariables: Record<string, codebuild.BuildEnvironmentVariable> =
  {
    DOMAIN_NAME: {
      type: codebuild.BuildEnvironmentVariableType.PARAMETER_STORE,
      value: "/findy-issuer-tool/domain-name",
    },
    SUB_DOMAIN_NAME: {
      type: codebuild.BuildEnvironmentVariableType.PARAMETER_STORE,
      value: "/findy-issuer-tool/sub-domain-name",
    },
    WALLET_DOMAIN_NAME: {
      type: codebuild.BuildEnvironmentVariableType.PARAMETER_STORE,
      value: "/findy-issuer-tool/wallet-domain-name",
    },
  };

export class InfraPipelineStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: InfraPipelineProperties) {
    super(scope, id, props);

    // Create configuration bucket for storing EB configuration files
    const confBucket = this.createConfigBucket(id, props.skipConfigCopy);

    // Create pipeline
    const pipeline = this.createPipeline(confBucket);

    // Add app to pipeline
    const deploy = new InfraPipelineStage(this, "Deploy", {
      env: props.env,
    });
    const deployStage = pipeline.addStage(deploy);

    // Add EB application update step
    deployStage.addPost(this.createEbUpdateStep(id, deploy));

    // Add frontend build step
    const frontBuildStep = this.createFrontendBuildStep();
    deployStage.addPost(frontBuildStep);

    // Add frontend deploy step
    deployStage.addPost(this.createFrontendDeployStep(frontBuildStep));

    // manually adjust logs retention
    pipeline.node.findAll().forEach((construct, index) => {
      if (construct instanceof codebuild.Project) {
        new logs.LogRetention(this, `LogRetention${index}`, {
          logGroupName: `/aws/codebuild/${construct.projectName}`,
          retention: logs.RetentionDays.ONE_MONTH,
        });
      }
    });
  }

  createConfigBucket(id: string, skipCopy: boolean = false) {
    const filesPath = "./.config";
    const confBucket = new Bucket(this, `${id}-conf-bucket`, {
      bucketName: `${id}-conf-bucket`.toLowerCase(),
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
    });

    const dockerRunJson = {
      AWSEBDockerrunVersion: "1",
      Image: {
        Name: "ghcr.io/findy-network/findy-issuer-tool",
        Update: "true",
      },
      Ports: [
        {
          ContainerPort: 3001,
          HostPort: 3001,
        },
      ],
    };
    const dockerRunPath = `${filesPath}/Dockerrun.aws.json`;
    mkdirSync(filesPath, { recursive: true });
    writeFileSync(dockerRunPath, JSON.stringify(dockerRunJson));
    execSync(`zip -r -j ${filesPath}/Dockerrun.zip ${dockerRunPath}`);
    execSync(`zip -r ${filesPath}/Dockerrun.zip .ebextensions`);

    if (!skipCopy) {
      new BucketDeployment(this, `${id}-bucket-deployment`, {
        sources: [Source.asset(filesPath)],
        destinationBucket: confBucket,
      });
    }
    return confBucket;
  }

  createPipeline(confBucket: IBucket) {
    const githubConnectionArn = StringParameter.valueForStringParameter(
      this,
      "/findy-issuer-tool/github-connection-arn"
    );

    const pipeline = new CodePipeline(this, "Pipeline", {
      pipelineName: "FindyIssuerToolPipeline",
      synth: new CodeBuildStep("SynthStep", {
        input: CodePipelineSource.connection(
          "findy-network/findy-issuer-tool",
          "master",
          {
            connectionArn: githubConnectionArn, // Created using the AWS console
          }
        ),
        installCommands: ["npm install -g aws-cdk"],
        buildEnvironment: {
          environmentVariables: {
            CDK_CONTEXT_JSON: {
              type: codebuild.BuildEnvironmentVariableType.PARAMETER_STORE,
              value: "/findy-issuer-tool/cdk-context",
            },
          },
        },
        commands: [
          "cd infra",
          `echo "$CDK_CONTEXT_JSON" > cdk.context.json`,
          "cat cdk.context.json",
          "npm ci",
          "npm run build",
          "npx cdk synth",
          "npm run pipeline:context",
        ],
        rolePolicyStatements: [
          new PolicyStatement({
            actions: ["ssm:PutParameter"],
            resources: [
              `arn:aws:ssm:${this.region}:${this.account}:parameter/findy-issuer-tool*`,
            ],
          }),
        ],
        primaryOutputDirectory: "infra/cdk.out",
      }),
      codeBuildDefaults: {
        buildEnvironment: {
          environmentVariables: {
            ...environmentVariables,
            CONF_BUCKET_NAME: {
              type: codebuild.BuildEnvironmentVariableType.PLAINTEXT,
              value: confBucket.bucketName,
            },
            CONF_BUCKET_ARN: {
              type: codebuild.BuildEnvironmentVariableType.PLAINTEXT,
              value: confBucket.bucketArn,
            },
          },
        },
      },
    });

    return pipeline;
  }

  createEbUpdateStep(id: string, deploy: InfraPipelineStage) {
    const deployRole = new Role(this, `${id}-deploy-role`, {
      assumedBy: new ServicePrincipal("codebuild.amazonaws.com"),
    });

    const ebPolicy = new PolicyStatement();
    ebPolicy.addActions(
      "elasticbeanstalk:*",
      "ec2:*",
      "ecr:*",
      "autoscaling:*",
      "cloudformation:*",
      "logs:*",
      "iam:PutRolePolicy",
      "s3:PutObject",
      "s3:PutObjectAcl",
      "s3:ListBucket",
      "s3:DeleteObject",
      "s3:Get*",
      "s3:CreateBucket"
    );
    ebPolicy.addResources("*");
    deployRole.addToPolicy(ebPolicy);

    return new CodeBuildStep("DeployBackendStep", {
      projectName: "DeployBackend",
      commands: [
        `VERSION="$(./infra/tools/version.sh ./api)-$(date +%s)"`,
        `aws elasticbeanstalk create-application-version --application-name $APP_NAME --version-label $VERSION --source-bundle S3Bucket=$CONF_BUCKET_NAME,S3Key=Dockerrun.zip`,
        `aws elasticbeanstalk update-environment --environment-name $ENV_NAME --version-label $VERSION`,
      ],
      envFromCfnOutputs: {
        APP_NAME: deploy.appName,
        ENV_NAME: deploy.envName,
      },
      role: deployRole,
    });
  }

  createFrontendBuildStep() {
    return new CodeBuildStep("BuildFrontendStep", {
      projectName: "BuildFrontend",
      commands: [
        "apk add bash",
        "cd frontend",
        "npm ci",
        "npm run build",
        "cd ..",
        "npm ci",
        "npm run licenses:report",
      ],
      buildEnvironment: {
        buildImage: codebuild.LinuxBuildImage.fromDockerRegistry(
          "node:16.14.2-alpine3.15"
        ),
        environmentVariables: {
          API_URL: {
            value: `https://${process.env.SUB_DOMAIN_NAME}.${process.env.DOMAIN_NAME}`,
          },
          DEFAULT_WEB_WALLET_LABEL: {
            value: "Findy Web Wallet",
          },
          DEFAULT_WEB_WALLET_URL: {
            value: `https://${process.env.WALLET_DOMAIN_NAME}/connect/`,
          },
        },
      },
      primaryOutputDirectory: "frontend/build",
    });
  }

  createFrontendDeployStep(frontBuildStep: CodeBuildStep) {
    return new CodeBuildStep("DeployFrontendStep", {
      input: frontBuildStep.primaryOutput,
      projectName: "DeployFrontend",
      commands: [
        `V1=$(curl https://$SUB_DOMAIN_NAME.$DOMAIN_NAME/version.txt)`,
        `V2=$(cat ./version.txt)`,
        `if [ "$V1" != "$V2" ]; then aws s3 sync --delete . s3://$SUB_DOMAIN_NAME.$DOMAIN_NAME; fi`,
      ],
      rolePolicyStatements: [
        new PolicyStatement({
          actions: ["s3:Put*", "s3:Delete*", "s3:Get*", "s3:List*"],
          resources: ["*"],
        }),
      ],
    });
  }
}
