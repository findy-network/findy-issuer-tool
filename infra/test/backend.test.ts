import * as cdk from "aws-cdk-lib";

import { Template } from "aws-cdk-lib/assertions";
import { Backend } from "../lib/backend";

test("Backend Created", () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, "MyTestStack", {
    env: { account: "123456789012", region: "us-east-1" },
  });

  new Backend(stack, "MyTestBackend", {
    account: stack.account,
  });

  const template = Template.fromStack(stack);

  template.resourceCountIs("AWS::ElasticBeanstalk::Application", 1);
  template.resourceCountIs("AWS::ElasticBeanstalk::ApplicationVersion", 1);
  template.resourceCountIs("AWS::ElasticBeanstalk::Environment", 1);
  template.resourceCountIs("AWS::DynamoDB::Table", 1);

  template.hasResourceProperties("AWS::ElasticBeanstalk::Application", {
    ApplicationName: "MyTestBackend-app",
    ResourceLifecycleConfig: {
      ServiceRole:
        "arn:aws:iam::123456789012:role/aws-service-role/elasticbeanstalk.amazonaws.com/AWSServiceRoleForElasticBeanstalk",
    },
  });

  expect(template).toMatchSnapshot();
});
