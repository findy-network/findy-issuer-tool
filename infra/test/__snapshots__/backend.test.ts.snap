// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Backend Created 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "MyTestBackendMyTestBackendapp73A3D5DA": Object {
      "Properties": Object {
        "ApplicationName": "MyTestBackend-app",
        "ResourceLifecycleConfig": Object {
          "ServiceRole": "arn:aws:iam::123456789012:role/aws-service-role/elasticbeanstalk.amazonaws.com/AWSServiceRoleForElasticBeanstalk",
          "VersionLifecycleConfig": Object {
            "MaxAgeRule": Object {
              "Enabled": true,
              "MaxAgeInDays": 3,
            },
          },
        },
      },
      "Type": "AWS::ElasticBeanstalk::Application",
    },
    "MyTestBackendMyTestBackendappversion8C6DDEFE": Object {
      "DependsOn": Array [
        "MyTestBackendMyTestBackendapp73A3D5DA",
      ],
      "Properties": Object {
        "ApplicationName": "MyTestBackend-app",
        "SourceBundle": Object {
          "S3Bucket": "conf-bucket",
          "S3Key": "Dockerrun.zip",
        },
      },
      "Type": "AWS::ElasticBeanstalk::ApplicationVersion",
    },
    "MyTestBackendMyTestBackendenvironmentDDB80453": Object {
      "Properties": Object {
        "ApplicationName": "MyTestBackend-app",
        "CNAMEPrefix": "issuer-tool",
        "EnvironmentName": "MyTestBackend-app",
        "OptionSettings": Array [
          Object {
            "Namespace": "aws:autoscaling:launchconfiguration",
            "OptionName": "IamInstanceProfile",
            "Value": Object {
              "Ref": "MyTestBackendissuertoolinstanceprofile0A99BA9E",
            },
          },
          Object {
            "Namespace": "aws:elasticbeanstalk:environment",
            "OptionName": "EnvironmentType",
            "Value": "SingleInstance",
          },
          Object {
            "Namespace": "aws:elasticbeanstalk:environment",
            "OptionName": "EnvironmentType",
            "Value": "SingleInstance",
          },
          Object {
            "Namespace": "aws:autoscaling:asg",
            "OptionName": "MinSize",
            "Value": "1",
          },
          Object {
            "Namespace": "aws:autoscaling:asg",
            "OptionName": "MaxSize",
            "Value": "1",
          },
        ],
        "SolutionStackName": "64bit Amazon Linux 2018.03 v2.17.7 running Docker 20.10.7-ce",
        "VersionLabel": Object {
          "Ref": "MyTestBackendMyTestBackendappversion8C6DDEFE",
        },
      },
      "Type": "AWS::ElasticBeanstalk::Environment",
    },
    "MyTestBackendMyTestBackendinstanceprofilerole3AC3276B": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ec2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": "dynamodb:*",
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::GetAtt": Array [
                      "MyTestBackendMyTestBackendtableBF3680E9",
                      "Arn",
                    ],
                  },
                  "Sid": "AllowDynamoDb",
                },
                Object {
                  "Action": Array [
                    "logs:CreateLogGroup",
                    "logs:CreateLogStream",
                    "logs:GetLogEvents",
                    "logs:PutLogEvents",
                    "logs:DescribeLogGroups",
                    "logs:DescribeLogStreams",
                    "logs:PutRetentionPolicy",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                  "Sid": "AllowCloudWatch",
                },
                Object {
                  "Action": "s3:GetObject",
                  "Effect": "Allow",
                  "Resource": "undefined/*",
                  "Sid": "AllowS3Read",
                },
                Object {
                  "Action": "secretsmanager:GetSecretValue",
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":secretsmanager:us-east-1:123456789012:secret:issuer-tool*",
                      ],
                    ],
                  },
                  "Sid": "AllowSecretRead",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "issuer-tool-policy",
          },
        ],
        "RoleName": "MyTestBackend-instance-profile-role",
      },
      "Type": "AWS::IAM::Role",
    },
    "MyTestBackendMyTestBackendtableBF3680E9": Object {
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "type",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "id",
            "AttributeType": "S",
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "type",
            "KeyType": "HASH",
          },
          Object {
            "AttributeName": "id",
            "KeyType": "RANGE",
          },
        ],
        "ProvisionedThroughput": Object {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
        "TableName": "issuerToolData",
      },
      "Type": "AWS::DynamoDB::Table",
    },
    "MyTestBackendissuertoolinstanceprofile0A99BA9E": Object {
      "Properties": Object {
        "InstanceProfileName": "MyTestBackend-instance-profile-role",
        "Path": "/",
        "Roles": Array [
          Object {
            "Ref": "MyTestBackendMyTestBackendinstanceprofilerole3AC3276B",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
